pkg/buildkit/worker/leasemanager.go

```diff
@@ -0,0 +1,345 @@
+	list, err := m.podClient.List(ctx, m.podListOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, pod := range list.Items {
+		if _, ok := pod.Annotations[leasedAnnotation]; !ok {
+			m.log.Info("Found unleased pod", "name", pod.Name, "namespace", pod.Namespace)
+			return &pod, nil
+		}
+	}
+
+	m.log.Info("No unleased pods found")
+	return nil, errNoUnleasedPods
+}
+
+func (m *podLeaseManager) buildURL(ctx context.Context, podName string) (string, error) {
+	m.log.Info("Querying for endpoints", "name", m.serviceName, "namespace", m.namespace)
+	endpoints, err := m.endpointsClient.Get(ctx, m.serviceName, metav1.GetOptions{})
+	if err != nil {
+		return "", err
+	}
+
+	var host string
+
+addressScan:
+	for _, subset := range endpoints.Subsets {
+		for _, address := range subset.Addresses {
+			if address.TargetRef.Name == podName {
+				for _, port := range subset.Ports {
+					if port.Port == m.servicePort {
+						host = strings.Join([]string{address.Hostname, endpoints.Name, endpoints.Namespace}, ".")
+						m.log.Info("Found eligible endpoint address", "host", host)
+
+						break addressScan
+					}
+				}
+			}
+		}
+	}
+
+	if host == "" {
+		return "", fmt.Errorf("endpoints %q does not expose pod %s on port %d", endpoints.Name, podName, 1234)
+	}
+
+	u, err := url.ParseRequestURI(fmt.Sprintf("tcp://%s:%d", host, m.servicePort))
+	if err != nil {
+		return "", fmt.Errorf("failed to parse endpoint url: %w", err)
+	}
+
+	return u.String(), nil
+}
+
+func (m *podLeaseManager) monitorPods(ctx context.Context) {
+	m.log.Info("Starting worker pod monitor", "syncTime", m.podSyncTime.String())
+	ticker := time.NewTicker(m.podSyncTime)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			m.log.Info("Shutting down worker pod monitor")
+			return
+		case <-ticker.C:
+			m.log.Info("Checking worker pool for expired leases")
+			if err := m.terminateUnleasedPods(ctx); err != nil {
+				m.log.Error(err, "Failed to clean worker pool")
+			}
+			m.log.Info("Worker pool lease check complete")
+		}
+	}
+}
+
+func (m *podLeaseManager) terminateUnleasedPods(ctx context.Context) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	m.log.Info("Querying for pods", "namespace", m.namespace, "opts", m.podListOptions)
+	list, err := m.podClient.List(ctx, m.podListOptions)
+	if err != nil {
+		return err
+	}
+
+	var removals []string
+	for idx := len(list.Items) - 1; idx >= 0; idx-- {
+		pod := list.Items[idx]
+		log := m.log.WithValues("podName", pod.Name)
+
+		if id, ok := pod.Annotations[managerIDAnnotation]; ok && id != m.uuid {
+			log.Info("Eligible for termination, manager id mismatch", "expected", m.uuid, "actual", id)
+			removals = append(removals, pod.Name)
+
+			continue
+		}
+
+		if _, leased := pod.Annotations[leasedAnnotation]; leased {
+			break
+		}
+
+		ts, ok := pod.Annotations[expiryTimeAnnotation]
+		if !ok {
+			m.log.Info("Eligible for termination, missing expiry time")
```

Comment:
May want to log the pod name (and probably in each of the log lines in the loop)?

Reply:
fixed in https://github.com/dominodatalab/hephaestus/pull/30/commits/7f6c17bdcd6db509f8ded5e5550c88729de22d36